digi/xbee/xsocket.py

```diff
@@ -0,0 +1,814 @@
+            ValueError: if the data to send is ``None``.
+            ValueError: if the number of bytes to send is ``0``.
+            XBeeException: if the connection with the XBee device is not open.
+            XBeeSocketException: if the socket is already open.
+            XBeeSocketException: if the send status is not ``SUCCESS``.
+        """
+        if data is None:
+            raise ValueError("Data to send cannot be None")
+        if len(data) == 0:
+            raise ValueError("The number of bytes to send must be at least 1")
+        if not self.__xbee_device.is_open():
+            raise XBeeException("XBee device must be open")
+        if self.__connected:
+            raise XBeeSocketException(message="Socket is already connected")
+
+        sent_bytes = 0
+
+        # If the socket is not created, create it first.
+        if self.__socket_id is None:
+            self.__create_socket()
+        # Send as many packets as needed to deliver all the provided data.
+        for chunk in self.__split_payload(data):
+            send_packet = SocketSendToPacket(self.__xbee_device.get_next_frame_id(), self.__socket_id,
+                                             IPv4Address(address[0]), address[1], chunk)
+            response_packet = self.__xbee_device.send_packet_sync_and_get_response(send_packet, self.__timeout)
+            self.__check_response(response_packet)
+            sent_bytes += len(chunk)
+        # Return the number of bytes sent.
+        return sent_bytes
+
+    def close(self):
+        """
+        Closes the socket.
+
+        Raises:
+            TimeoutException: if the close response is not received in the configured timeout.
+            XBeeException: if the connection with the XBee device is not open.
+            XBeeSocketException: if the close status is not ``SUCCESS``.
+        """
+        if self.__socket_id is None or (not self.__connected and not self.__source_port):
+            return
+        if not self.__xbee_device.is_open():
+            raise XBeeException("XBee device must be open")
+
+        close_packet = SocketClosePacket(self.__xbee_device.get_next_frame_id(), self.__socket_id)
+        response_packet = self.__xbee_device.send_packet_sync_and_get_response(close_packet, self.__timeout)
+        self.__check_response(response_packet)
+
+        self.__connected = False
+        self.__socket_id = None
+        self.__source_port = None
+        self.__data_received = bytearray()
+        self.__data_received_from_dict = OrderedDict()
+        self.__unregister_state_callback()
+        self.__unregister_data_received_callback()
+        self.__unregister_data_received_from_callback()
+
+    def setsocketopt(self, option, value):
+        """
+        Sets the value of the given socket option.
+
+        Args:
+            option (:class:`.SocketOption`): the socket option to set its value.
+            value (Bytearray): the new value of the socket option.
```

Comment:
What values are allowed to be set here? If they are a fixed number of values, maybe we should consider creating an enumeration and then check if the given value matches a valid one, if not, return with an exception.